spyder_kernels/comms/commbase.py

```diff
@@ -0,0 +1,372 @@
+        content: dict
+            The (JSONable) content of the message
+        data: any
+            Any object that is serializable by cloudpickle (should be most
+            things). Will arrive as cloudpickled bytes in `.buffers[0]`.
+        """
+        if not self.is_open():
+            raise CommError("The comm is not connected.")
+        import cloudpickle
+        msg_dict = {
+            'spyder_msg_type': spyder_msg_type,
+            'content': content,
+            }
+        buffers = [cloudpickle.dumps(data, protocol=PICKLE_PROTOCOL)]
+        self._comm.send(msg_dict, buffers=buffers)
+
+    @property
+    def _comm_name(self):
+        """
+        Get the name used for the underlying comms.
+        """
+        return 'spyder_api'
+
+    def _register_message_handler(self, message_id, handler):
+        """
+        Register a message handler.
+
+        Parameters
+        ----------
+        message_id : str
+            The identifier for the message
+        handler : callback
+            A function to handle the message. This is called with 3 arguments:
+                - msg_dict: A dictionary with message information.
+                - buffer: The data transmitted in the buffer
+                - load_exception: Exception from buffer unpickling
+            Pass None to unregister the message_id
+        """
+        if handler is None:
+            self._message_handlers.pop(message_id, None)
+            return
+
+        self._message_handlers[message_id] = handler
+
+    def _register_comm(self, comm):
+        """
+        Open a new comm to the kernel.
+        """
+        if self._comm:
+            raise CommError("Close the comm first.")
+        comm.on_msg(self._comm_message)
+        comm.on_close(self._comm_close)
+        self._comm = comm
+
+    def _comm_close(self, msg):
+        """Close comm."""
+        self._comm = None
+
+    def _comm_message(self, msg):
+        """
+        Handle internal spyder messages.
+        """
+        # Load the buffer. Only one is supported.
+        try:
+            if PY2:
+                buffer = cloudpickle.loads(msg['buffers'][0])
+            else:
+                buffer = cloudpickle.loads(bytes(msg['buffers'][0]))
+            load_exception = None
+        except Exception as e:
+            load_exception = e
+            buffer = None
+
+        # Get message dict
+        msg_dict = msg['content']['data']
+
+        spyder_msg_type = msg_dict['spyder_msg_type']
+
+        if spyder_msg_type in self._message_handlers:
+            self._message_handlers[spyder_msg_type](
+                msg_dict, buffer, load_exception)
+            return
+
+        logger.debug("No such spyder message type: %s" % spyder_msg_type)
+
+    def _handle_remote_call(self, msg, buffer, load_exception):
+        """Handle a remote call."""
+        msg_dict = msg['content']
+        if load_exception:
+            raise load_exception
```

Comment:
From #112:

> If something doesn't happen, it can be frustrating for the user. As a user, I would rather know that something went wrong and be able to report it rather than not understanding why something is not happening.

I understand that, but then we should display the exception when Spyder is running in debug mode (so we'd to inform the kernel about that). Or reroute these exceptions to Spyder and show them through our error report dialog.

Otherwise I doubt people (other than Spyder developers, I mean) would be able to make anything out of these exceptions.